### ISSUE TITLE
Password support for SSL certificates

### ISSUE THREAD

**Issue by Dobatymo** [cherrypy/cherrypy#1583]
* **I'm submitting a ...**
[ ] bug report
[x] feature request
[ ] question about the decisions made in the repository

`SSLContext.load_cert_chain(certfile, keyfile=None, password=None) supports a password argument` (new in python 3.3) to use for encrypted private keys. Please make this available in `CherryPyWSGIServer`.

#1165 related but not relevant anymore (i think)

**Comment by webknjaz** [comment]
This is similar to to recent changes (0aa622f8ec930c95e45455acae810e23f3c47105, ref cherrypy/cheroot#22).
I'm not sure about having time for this, but a PR from you would be a good start :)

**Comment by webknjaz** [comment]
*UPD:* The Cheroot side of the implementation https://github.com/cherrypy/cheroot/pull/752 has been merged and released under v11.1.0. Now what's left is to have a CherryPy PR surfacing the new configuration setting `private_key_password`.

**Comment by webknjaz** [comment]
@jatalahd will you be taking this up?

**Comment by jatalahd** [comment]
@webknjaz Yes, of course I need to go all the way with this. Sorry about the delay. There is now a pull request available with minimal changes. If more is needed please let me know. With these changes, there are no dependency conflicts with older cheroot versions, that do not have the private_key_password option in the ssl adapters.

Also when testing this, I noticed that the pyopenssl adapter does not work anymore in the old and documented workflow, where the password is entered manually in case of a password protected private key. Currently the password callback is triggered also when `private_key_password = None`. I will make a related pull request to cheroot side that fixes this issue. (see pull request: https://github.com/cherrypy/cheroot/pull/798)

**Comment by webknjaz** [comment]
> Also when testing this, I noticed that the pyopenssl adapter does not work anymore in the old and documented workflow, where the password is entered manually in case of a password protected private key. Currently the password callback is triggered also when `private_key_password = None`.

Could you shed more light on what you mean by this difference in behavior? I've also posted comments on that PR since I don't fully understand why you think is broken.

**Comment by jatalahd** [comment]
@webknjaz The feature that I am trying to preserve is written in the documentation:
https://docs.cherrypy.dev/en/latest/deploy.html#ssl-support

> Protecting your private key with a password is much more secure, but requires that you enter the password every time you use the key. For example, you may have to enter the password when you start or restart your CherryPy server.

So, if someone has been using the protected private key this way until now, manually entering the password, update to latest cherrypy (with the current state of development) when using pyopenssl adapter it does not ask user to input the password, but fails due to the default password `b''` not matching the password of the key. The builtin adapter still preserves the old functionality so that users will not notice anything different when updating to the current development version.

When I am proposing to skip calling the callback method in pyopenssl adapter in the case when password is not provided, then the old functionality remains as documented and behaves the same as the builtin adapter does now.

When pushing the first versions of pytests in the cheroot repo for the adapter changes, I tried to make a unit test for the "manual" input of the password, but never got it working. I know for sure that at least in Windows environment it is impossible to make such a unit test.

**Comment by webknjaz** [comment]
> [@webknjaz](https://github.com/webknjaz) The feature that I am trying to preserve is written in the documentation: https://docs.cherrypy.dev/en/latest/deploy.html#ssl-support
> 
> > Protecting your private key with a password is much more secure, but requires that you enter the password every time you use the key. For example, you may have to enter the password when you start or restart your CherryPy server.

Ah, I see. Still, I'd fact-check that first. I'm not aware of anything across Cheroot / CherryPy actually performing any interactive password dialog. That text was written long ago. I'm under impression that the context there is that something external was expected to be deciphering the key.

> So, if someone has been using the protected private key this way until now, manually entering the password, update to latest cherrypy (with the current state of development) when using pyopenssl adapter it does not ask user to input the password, but fails due to the default password `b''` not matching the password of the key. The builtin adapter still preserves the old functionality so that users will not notice anything different when updating to the current development version.
> 
> When I am proposing to skip calling the callback method in pyopenssl adapter in the case when password is not provided, then the old functionality remains as documented and behaves the same as the builtin adapter does now.
> 
> When pushing the first versions of pytests in the cheroot repo for the adapter changes, I tried to make a unit test for the "manual" input of the password, but never got it working. I know for sure that at least in Windows environment it is impossible to make such a unit test.

Have you actually tried using an encrypted key w/o a password w/ pyopenssl? Does it really work like the doc says. It might be that you can't create a test because it doesn't work.
